rholang/src/main/scala/coop/rchain/rholang/interpreter/compiler/normalize.scala

```diff
@@ -1061,17 +1191,222 @@ object ProcNormalizeMatcher {
                       s"Illegal top level connective in bundle at position: line: ${b.line_num}, column: ${b.col_num}."
                     )
                   )
-                } else if (targetResult.knownFree.wildcards.nonEmpty || targetResult.knownFree.levelBindings.nonEmpty) {
+                } else if (targetResult.freeMap.wildcards.nonEmpty || targetResult.freeMap.levelBindings.nonEmpty) {
                   error(targetResult)
                 } else {
                   val newBundle: Bundle = targetResult.par.singleBundle() match {
                     case Some(single) => outermostBundle.merge(single)
                     case None         => outermostBundle
                   }
-                  ProcVisitOutputs(input.par.prepend(newBundle), input.knownFree).pure[M]
+                  ProcVisitOutputs(input.par.prepend(newBundle), input.freeMap).pure[M]
                 }
         } yield res
 
+      case p: PLet => {
+
+        p.decls_ match {
+
+          case concDeclsImpl: ConcDeclsImpl =>
+            def extractNamesAndProcs(decl: Decl): (ListName, NameRemainder, ListProc) =
+              decl match {
+                case declImpl: DeclImpl =>
+                  (declImpl.listname_, declImpl.nameremainder_, declImpl.listproc_)
+              }
+
+            val (listNames, listNameRemainders, listProcs) =
+              (extractNamesAndProcs(p.decl_) :: concDeclsImpl.listconcdecl_.toList.map {
+                case concDeclImpl: ConcDeclImpl => extractNamesAndProcs(concDeclImpl.decl_)
+              }).unzip3
+
+            /*
+             It is not necessary to use UUIDs to achieve concurrent let declarations.
+             While there is the possibility for collisions with either variables declared by the user
+             or variables declared within this translation, the chances for collision are astronomically
+             small (see analysis here: https://towardsdatascience.com/are-uuids-really-unique-57eb80fc2a87).
+             A strictly correct approach would be one that performs a ADT rather than an AST translation, which
+             was not done here due to time constraints.
+             */
+            val variableNames = List.fill(listNames.size)(UUID.randomUUID().toString)
+
+            val psends = variableNames.zip(listProcs).map {
+              case (variableName, listProc) =>
+                new PSend(new NameVar(variableName), new SendSingle(), listProc)
```

Comment:
The `NameVar` is not something new which this pr introduced. 
I am curious how codes before this pr work with `variableName` in NameVar.
I mean instead of `UUID`, what does these `variableName` come from?

Reply:
The variable name is supplied by the programmer in the source code.